src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+        Int counter = new Int(oldCell.smokeIndicator());
+        return NeighbourUtils.neighboursStream(cellIndex)
+                .filter(oldState::hasCell)
+                .filter(neighbourIndex -> oldState.getCell(neighbourIndex).isFluid())
+                .map(neighbourIndex -> {
+                    double diff = 0.0;
+                    if (NeighbourUtils.up(cellIndex).equals(neighbourIndex)) {
+                        diff = -Math.min(oldCell.smokeIndicator(), 100 - oldState.getCell(neighbourIndex).smokeIndicator());
+                        counter.v += diff;
+                    } else if (NeighbourUtils.down(cellIndex).equals(neighbourIndex)) {
+                        diff = Math.min(oldState.getCell(neighbourIndex).smokeIndicator(),
+                                100 - oldCell.smokeIndicator());
+                        counter.v += diff;
+                    } else if (isHorizontal(cellIndex, neighbourIndex)) {
+                        int numberOfNeighboursMine = N(oldState, cellIndex).size() + 1;
+                        int numberOfNeighboursHim = N(oldState, neighbourIndex).size() + 1;
+                        if (cellAboveCanNotTakeSmoke(oldState, cellIndex)) {
+                            diff -= Math.min((100 - oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim,
+                                    (counter.v) / numberOfNeighboursMine);
+                        }
+                        if (cellAboveCanNotTakeSmoke(oldState, neighbourIndex)) {
+                            diff += Math.min((100 - counter.v) / numberOfNeighboursMine,
+                                    (oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim);
+                        }
+                    }
+                    return diff;
+                }).mapToDouble(Double::doubleValue).sum();
+    }
+
+    private boolean cellAboveCanNotTakeSmoke(State state, Vector3i index) {
+        Vector3i cellAbove = NeighbourUtils.up(index);
+        return (state.hasCell(cellAbove) && state.getCell(cellAbove).smokeIndicator() >= 50) ||
+                (state.hasCell(cellAbove) && !state.getCell(cellAbove).isFluid()) ||
+                !state.hasCell(cellAbove);
+    }
+
+    private List<Cell> N(State state, Vector3i cellIndex) {
+        List<Cell> res = new ArrayList<>();
+        NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.X)
+                .filter(state::hasCell)
+                .filter(index -> state.getCell(index).smokeIndicator() < 100)
+                .forEach(index -> res.add(state.getCell(index)));
+        NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.Z)
+                .filter(state::hasCell)
+                .filter(index -> state.getCell(index).smokeIndicator() < 100)
+                .forEach(index -> res.add(state.getCell(index)));
+        return res;
+    }
+
+    private boolean isHorizontal(Vector3i cellIndex, Vector3i neighbourIndex) {
```

Comment:
We won't need that method, because befor I have shown that the condition for horizontal neighbours will always be true.